### PR TITLE
PR from work_branch to main branch

### PR DESCRIPTION
**Pull request by JofielB**
## Summary
Initialize the Flight project and create the CRUD for the Airport entity

## Changes
- Initialize the project 
- Create the entity for the airports
- Create the repository of the airport entity
- Create the controller(CRUD) of the airport entity
- Create a custom exception that is used in the `AirportController`

## Target
`main`

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -36,6 +36,12 @@
 			<artifactId>spring-boot-starter-test</artifactId>
 			<scope>test</scope>
 		</dependency>
+		<dependency>
+			<groupId>org.projectlombok</groupId>
+			<artifactId>lombok</artifactId>
```
Comment (janvmusic) [inline pom.xml:41]: What is the advantage of having this in the project?

File: src/main/java/com/example/flight/FlightApplication.java [hunk 2]
```diff
@@ -1,13 +1,12 @@
...
+package com.example.flight;
 
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 
 @SpringBootApplication
 public class FlightApplication {
 
-	public static void main(String[] args) {
+	public static void main(String... args) {
```
Comment (janvmusic) [inline src/main/java/com/example/flight/FlightApplication.java:9]: Why this change?

File: src/main/java/com/example/flight/controllers/AirportController.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+class AirportController {
+
+    private final AirportRepository repository;
```
Comment (janvmusic) [inline src/main/java/com/example/flight/controllers/AirportController.java]: Controllers should not talk directly to the repository. What do you think about this?

File: src/main/java/com/example/flight/controllers/AirportController.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    private AirportController(AirportRepository repository) {
+        this.repository = repository;
+    }
+
+    @GetMapping("/test")
```
Comment (janvmusic) [inline src/main/java/com/example/flight/controllers/AirportController.java]: Not needed, remove

File: src/main/java/com/example/flight/controllers/AirportController.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        return "hello world";
+    }
+
+    @GetMapping("/airports")
+    public List<Airport> all() {
+        return repository.findAll();
+    }
+
+    @PostMapping("/airports")
+    public Airport newEmployee(@RequestBody Airport newAirport) {
```
Comment (janvmusic) [inline src/main/java/com/example/flight/controllers/AirportController.java]: Incorrect name... `newEmployee`?

Beware of copy paste

File: src/main/java/com/example/flight/controllers/AirportController.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        return repository.save(newAirport);
+    }
+
+    @GetMapping("/airports/{id}")
+    public Airport one(@PathVariable Long id) {
```
Comment (janvmusic) [inline src/main/java/com/example/flight/controllers/AirportController.java]: What's the meaning of `one`? One what?

Try to use a convention for these cases... maybe {Get | Post | Put | Delete}{methodName} for this example I would use `getAirportById`

File: src/main/java/com/example/flight/controllers/AirportController.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+        return repository.findById(id)
```
Comment (janvmusic) [inline src/main/java/com/example/flight/controllers/AirportController.java]: If I recall correctly you could use `AspectProgramming` and create a controller exception helper to intercept these scenarios

File: src/main/java/com/example/flight/controllers/AirportController.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                .orElseThrow(() -> new AirportNotFoundException(id));
+    }
+
+    @PutMapping("/airports/{id}")
+    public Airport replaceEmployee(@RequestBody Airport newEmployee, @PathVariable Long id) {
```
Comment (janvmusic) [inline src/main/java/com/example/flight/controllers/AirportController.java]: Wrong name `replaceEmployeeName`

File: src/main/java/com/example/flight/controllers/AirportController.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @GetMapping("/airports")
+    public List<Airport> all() {
+        return repository.findAll();
+    }
+
+    @PostMapping("/airports")
+    public Airport newEmployee(@RequestBody Airport newAirport) {
+        return repository.save(newAirport);
```
Comment (janvmusic) [inline src/main/java/com/example/flight/controllers/AirportController.java]: Do you think you should expose this to other users to add airports? How often a new airport is created?

File: src/main/java/com/example/flight/controllers/AirportController.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+        return repository.findById(id)
+                .orElseThrow(() -> new AirportNotFoundException(id));
+    }
+
+    @PutMapping("/airports/{id}")
+    public Airport replaceEmployee(@RequestBody Airport newEmployee, @PathVariable Long id) {
+
+        return repository.findById(id)
+                .map(employee -> {
```
Comment (janvmusic) [inline src/main/java/com/example/flight/controllers/AirportController.java]: Why are we loading `employees`?

File: src/main/java/com/example/flight/controllers/AirportController.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+                    employee.set_name(newEmployee.get_name());
```
Comment (janvmusic) [inline src/main/java/com/example/flight/controllers/AirportController.java]: Weird pattern `get_name` use camel case

File: src/main/java/com/example/flight/controllers/AirportController.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    public Airport one(@PathVariable Long id) {
+
+        return repository.findById(id)
+                .orElseThrow(() -> new AirportNotFoundException(id));
+    }
+
+    @PutMapping("/airports/{id}")
+    public Airport replaceEmployee(@RequestBody Airport newEmployee, @PathVariable Long id) {
+
+        return repository.findById(id)
```
Comment (janvmusic) [inline src/main/java/com/example/flight/controllers/AirportController.java]: It feels that this is not Controller purpose I would change this to a service or leave it to the repository

File: src/main/java/com/example/flight/controllers/AirportController.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+                .map(employee -> {
+                    employee.set_name(newEmployee.get_name());
+                    employee.set_airportCode(newEmployee.get_airportCode());
+                    return repository.save(employee);
+                })
+                .orElseGet(() -> {
+                    newEmployee.setId(id);
```
Comment (janvmusic) [inline src/main/java/com/example/flight/controllers/AirportController.java]: So if I can't get an airport, then I should store it as new?

Let's discuss

**Review by janvmusic — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 1 above)
Comment (JofielB) [reply pom.xml:41]: The advantage is having cleaner POJOs and also, I wanted to practice using that library.

File: src/main/java/com/example/flight/FlightApplication.java (hunk 2 above)
Comment (JofielB) [reply src/main/java/com/example/flight/FlightApplication.java:9]: Actually, there isn't a particular reason for that change, I just like the three dots

File: src/main/java/com/example/flight/controllers/AirportController.java (hunk 4 above)
Comment (JofielB) [reply src/main/java/com/example/flight/controllers/AirportController.java]: Removed

File: src/main/java/com/example/flight/controllers/AirportController.java (hunk 5 above)
Comment (JofielB) [reply src/main/java/com/example/flight/controllers/AirportController.java]: Name changed

File: src/main/java/com/example/flight/controllers/AirportController.java (hunk 6 above)
Comment (JofielB) [reply src/main/java/com/example/flight/controllers/AirportController.java]: Names changed

File: src/main/java/com/example/flight/controllers/AirportController.java (hunk 8 above)
Comment (JofielB) [reply src/main/java/com/example/flight/controllers/AirportController.java]: Name changed

File: pom.xml (hunk 1 above)
Comment (janvmusic) [reply pom.xml:41]: The second reason is good, however for small projects might be overkiller.

File: src/main/java/com/example/flight/controllers/AirportController.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+        this.service = service;
+    }
+
+    @GetMapping
+    public List<Airport> getAllAirports() {
+        return service.getAllAirports();
+    }
+
+    @PostMapping
+    public Airport postAirport(@RequestBody Airport airport) {
```
Comment (janvmusic) [inline src/main/java/com/example/flight/controllers/AirportController.java]: I know that I suggested `actionMethodName` but I prefer readability for this case.

Which is `createAirport`, however, my question remains. Why would a user create an airport??
